modules/ml/src/main/java/org/apache/ignite/ml/trainers/TrainerTransformers.java

```diff
@@ -0,0 +1,282 @@
+        DatasetBuilder<K, V> datasetBuilder,
+        int ensembleSize,
+        double subsampleRatio,
+        int featuresVectorSize,
+        int maximumFeaturesCntPerMdl,
+        IgniteBiFunction<K, V, Vector> extractor,
+        PredictionsAggregator aggregator,
+        LearningEnvironment environment) {
+        MLLogger log = environment.logger(datasetBuilder.getClass());
+        log.log(MLLogger.VerboseLevel.LOW, "Start learning.");
+
+        List<Map<Integer, Integer>> mappings = null;
+        if (featuresVectorSize > 0) {
+            mappings = IntStream.range(0, ensembleSize).mapToObj(modelIdx -> getMapping(featuresVectorSize, maximumFeaturesCntPerMdl))
+                .collect(Collectors.toList());
+        }
+
+        Long startTs = System.currentTimeMillis();
+
+        DatasetBuilder<K, V> bootstrappedBuilder = datasetBuilder.addStreamTransformer(
+            // Sequentiality of stream here is needed because we use instance of
+            // RNG as data, to make it deterministic we should fix order.
+            (s, rnd) -> s.sequential().flatMap(en -> repeatEntry(en, rnd)),
+            () -> new PoissonDistribution(subsampleRatio)
+        );
+
+        List<IgniteSupplier<M>> tasks = new ArrayList<>();
+        List<IgniteBiFunction<K, V, Vector>> extractors = null;
+        if (mappings != null) {
+            extractors = mappings
+                .stream()
+                .map(m -> wrapExtractor(extractor, m))
+                .collect(Collectors.toList());
+        }
+
+        for (int i = 0; i < ensembleSize; i++) {
+            tasks.add(trainingTaskGenerator.apply(bootstrappedBuilder, i, mappings != null ? extractors.get(i) : extractor));
+        }
+
+        List<ModelWithEndomorphism<Vector, Double, M>> models = environment.parallelismStrategy().submit(tasks)
+            .stream()
+            .map(Promise::unsafeGet)
+            .map(ModelWithEndomorphism<Vector, Double, M>::new)
+            .collect(Collectors.toList());
+
+        // If we need to do projection, do it.
+        if (mappings != null) {
+            for (int i = 0; i < models.size(); i++) {
+                models.get(i).setEndo(getProjector(mappings.get(i)));
+            }
+        }
+
+        double learningTime = (double) (System.currentTimeMillis() - startTs) / 1000.0;
+        log.log(MLLogger.VerboseLevel.LOW, "The training time was %.2fs.", learningTime);
+        log.log(MLLogger.VerboseLevel.LOW, "Learning finished.");
+
+        return new ModelsComposition(models, aggregator);
+    }
+
+    @NotNull
+    public static Map<Integer, Integer> getMapping(int featuresVectorSize, int maximumFeaturesCntPerMdl) {
+        int[] featureIdxs = Utils.selectKDistinct(featuresVectorSize, maximumFeaturesCntPerMdl, new Random());
+        Map<Integer, Integer> featureMapping = new HashMap<>();
+        IntStream.range(0, maximumFeaturesCntPerMdl)
+            .forEach(localId -> featureMapping.put(localId, featureIdxs[localId]));
```

Comment:
Do I understand correctly that `featureMapping` and `featureIdxs` are identical after this line?